🚀 1단계 - 예약 기능 구현

junseong94 opened the pull request:
안녕하세요 리뷰어님

1-1단계 완료 및 1-2 DB 설정 완료하여 PR 드립니다.

앞으로도 잘부탁드립니다! :)

junseong94 committed:
실습 페이지 설정 완료

junseong94 committed:
예약 관리 API - 예약조회 API 구현 완료

junseong94 committed:
예약 관리 API - 예약추가 API 구현 완료

junseong94 committed:
예약 관리 API - 예약취소 API 구현 완료

junseong94 committed:
h2 database 설정 완료

Hyeon9mak commented on src/main/java/roomescape/controller/Reservation.java:
```diff
@@ -0,0 +1,39 @@
+package roomescape.controller;
+
+public class Reservation {
+
+    private Long id;
+    private String name;
+    private String date;
+    private String time;
+
+    public Reservation(Long id, String name, String date, String time) {
+        this.id = id;
+        this.name = name;
+        this.date = date;
+        this.time = time;
+    }
+
+    public ReservationDto toDto() {
+        return new ReservationDto(name, date, time);
+    }
```
`Reservation` 은 비즈니스 로직을 수행하는 중요한 객체가 될 것으로 예상되는데요,
반면 `ReservationDto` 는 기획, client 의 요청에 따라 수시로 모양이 변경될 객체가 될 것 같아요.

현재 `Reservation` -> `ReservationDto` 의존 방향을 갖고 있는데,
이런 경우 `ReservationDto` 의 변경에 `Reservation` 이 계속해서 영향을 받게 될 것 같아요.
`Reservation` 이 잦은 변경으로부터 영향을 받지 않도록 의존 방향을 없애거나 제거하는 건 어떨까요?

Hyeon9mak commented on src/main/java/roomescape/controller/Reservation.java:
```diff
@@ -0,0 +1,39 @@
+        return time;
+    }
+
+    public Reservation() {
+    }
```
기본 생성자가 없어도 충분하겠는데요? 😉

Hyeon9mak commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+
+@RestController
+public class ReservationController {
+
+    private AtomicLong index = new AtomicLong();
+    private List<Reservation> reservations = List.of(
+            new Reservation(index.incrementAndGet(), "브라운", "2023-08-05", "15:40"),
+            new Reservation(index.incrementAndGet(), "코니", "2023-08-05", "15:40"));
```
동시성 제어를 위한 AtomicLong 활용! 👍 
시작 index 값이 무엇인지 `new AtomicLong(1)` 등으로 명세해주면 가독성이 더욱 높아지겠네요!

Hyeon9mak replied:
인덱스를 통해 값을 조회하므로... List 대신 더 잘어울리는 자료구조가 있을 것 같은데요? 😉 
(hint: key, value)

Hyeon9mak commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,41 @@
+        return ResponseEntity.ok().body(reservation.toDto());
+    }
+
+    @DeleteMapping("/reservations/{id}")
+    public ResponseEntity deleteReservation(@PathVariable Long id) {
+        reservations = reservations.stream()
+                .filter(reservation -> !reservation.getId().equals(id))
+                .toList();
+        return ResponseEntity.ok().build();
+    }
```
id 에 해당되는 원소가 존재하지 않는 경우는 어떻게 될까요?
사용자가 이를 인지할 수 있도록 해주면 더욱 좋겠네요!

Hyeon9mak commented on src/main/java/roomescape/controller/ReservationDto.java:
```diff
@@ -0,0 +1,37 @@
+        return time;
+    }
+
+    public Reservation toEntity() {
+        final AtomicLong atomicLong = new AtomicLong();
+        return new Reservation(atomicLong.incrementAndGet(), name, date, time);
+    }
```
id 생성 주체가 이곳저곳 존재하기 때문에, 자칫 중복 id 가 생성될 수도 있겠어요.
id 생성 주체를 한 곳으로 통일 시키는 것은 어떨까요?

Hyeon9mak replied:
`ReservationDto` 가 기본적으로 id 를 갖고 있어서, 생성시 문제가 발생할 수도 있겠는데요,
요청 방향 dto 와 응답 방향 dto 를 분리하면 id 필드에 대한 부담이 사라지겠어요!

Hyeon9mak commented on src/test/java/roomescape/controller/ReservationControllerTest.java:
```diff
@@ -0,0 +1,74 @@
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+class ReservationControllerTest {
+
+    @Test
+    void readReservation() {
+        var response = RestAssured
+                .given().log().all()
+                .when().get("/reservations")
+                .then().log().all().extract();
+
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.OK.value());
+        assertThat(response.jsonPath().getList(".", ReservationDto.class)).hasSize(2);
+    }
+
```
`@DisplayName` 으로 어떤 테스트인지 명세해주면 더욱 좋겠네요 😄

Hyeon9mak reviewed (changes requested):
안녕하세요 준성님! 방탈출 예약 관리 미션을 함께할 최현구입니다. 🙇‍♂️ 
DDD 미션에 이어서 뵙네요 😄 

예약기능 구현 잘 해주셨습니다! 몇 가지 코멘트 남겼으니 확인해주세요.
궁금하신 점 얼마든지 질문 남겨주세요~

junseong94 committed:
refactor : Reservation, ReservationController
create : ReservationResponseDto
update : ReservationDto to ReservationRequestDto

Hyeon9mak commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,63 @@
+import java.util.HashMap;
+import java.util.List;
+
+@RestController
+public class ReservationController {
+
+    private static final HashMap<Long, Reservation> reservations = new HashMap<>();
+    private final Reservation reservation1 = new Reservation("제이슨", "2023-08-05", "15:40");
+    private final Reservation reservation2 = new Reservation("심슨", "2023-08-05", "15:40");
+
```
`@RestController` 는 싱글톤을 보장하죠.
그렇다면 static 하게 선언하는 것과 아닌 것의 차이는?!

Hyeon9mak replied:
구현체는 `HashMap` 를 사용해야하지만, 명세는 `Map` 인터페이스를 사용해도 되겠네요!

```java
Map<Long, Reservation> reservations = new HashMap<>();
```

Hyeon9mak commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,63 @@
+        return ResponseEntity.ok().body(responseDto);
+    }
+
+    @DeleteMapping("/reservations/{id}")
+    public ResponseEntity deleteReservation(@PathVariable Long id) {
+        if (!reservations.containsKey(id)) {
+            return ResponseEntity.notFound().build();
+        } else {
```
😉 👍

Hyeon9mak commented on src/test/java/roomescape/controller/ReservationControllerTest.java:
```diff
@@ -0,0 +1,76 @@
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+class ReservationControllerTest {
+
+    @DisplayName("전체 예약을 조회 합니다.")
+    @Test
+    void readReservation() {
```
👍

Hyeon9mak reviewed (approved):
안녕하세요 준성님! 피드백 반영 잘해주셨습니다 👍 
아주 간단한 코멘트 하나 더 남겼습니다!
다음 단계 진행에 참고해주세요 😄
